pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java

```diff
@@ -0,0 +1,137 @@
+  private static final int MAX_NUM_BYTES_PER_VALUE = 512;
+  private static final int NUM_READS = 200_000;
+  private static final long RANDOM_SEED = System.currentTimeMillis();
+  private static final Random RANDOM = new Random(RANDOM_SEED);
+
+  private final PinotDataBufferMemoryManager _memoryManager = new DirectMemoryManager("testSegment");
+  private final byte[][] _values = new byte[NUM_VALUES][];
+
+  @BeforeClass
+  public void setUp() {
+    System.out.println("Random seed: " + RANDOM_SEED);
+
+    for (int i = 0; i < NUM_VALUES; i++) {
+      int numBytes = RANDOM.nextInt(MAX_NUM_BYTES_PER_VALUE);
+      byte[] value = new byte[numBytes];
+      RANDOM.nextBytes(value);
+      _values[i] = value;
+    }
+  }
+
+  @Test
+  public void testAdd()
+      throws Exception {
+    try (OffHeapMutableBytesStore offHeapMutableBytesStore = new OffHeapMutableBytesStore(_memoryManager, null)) {
+      for (int i = 0; i < NUM_VALUES; i++) {
+        assertEquals(offHeapMutableBytesStore.add(_values[i]), i);
+      }
+    }
+  }
+
+  @Test
+  public void testGet()
+      throws Exception {
+    try (OffHeapMutableBytesStore offHeapMutableBytesStore = new OffHeapMutableBytesStore(_memoryManager, null)) {
+      for (int i = 0; i < NUM_VALUES; i++) {
+        offHeapMutableBytesStore.add(_values[i]);
+      }
+      for (int i = 0; i < NUM_READS; i++) {
+        int index = RANDOM.nextInt(NUM_VALUES);
+        assertTrue(Arrays.equals(offHeapMutableBytesStore.get(index), _values[index]));
+      }
+    }
+  }
+
+  @Test
+  public void testEqualsValueAt()
+      throws Exception {
+    try (OffHeapMutableBytesStore offHeapMutableBytesStore = new OffHeapMutableBytesStore(_memoryManager, null)) {
+      for (int i = 0; i < NUM_VALUES; i++) {
+        offHeapMutableBytesStore.add(_values[i]);
+      }
+      for (int i = 0; i < NUM_READS; i++) {
+        int index = RANDOM.nextInt(NUM_VALUES);
+        assertTrue(offHeapMutableBytesStore.equalsValueAt(index, _values[index]));
+        if (!Arrays.equals(_values[index], _values[0])) {
+          assertFalse(offHeapMutableBytesStore.equalsValueAt(0, _values[index]));
+          assertFalse(offHeapMutableBytesStore.equalsValueAt(index, _values[0]));
+        }
+      }
+    }
+  }
+
+  @Test
+  public void concurrentReadWrite()
+      throws Exception {
+    int numReaders = 4;
+    ExecutorService executorService = Executors.newFixedThreadPool(numReaders + 1);
+    List<Future> futures = new ArrayList<>(numReaders + 1);
+    try (OffHeapMutableBytesStore offHeapMutableBytesStore = new OffHeapMutableBytesStore(_memoryManager, null)) {
+      // Writer thread
+      futures.add(executorService.submit(() -> {
+        for (int i = 0; i < NUM_VALUES; i++) {
+          offHeapMutableBytesStore.add(_values[i]);
+        }
+      }));
+
+      // Reader threads
+      for (int i = 0; i < numReaders; i++) {
+        futures.add(executorService.submit(() -> {
+          for (int j = 0; j < NUM_READS; j++) {
```

Comment:
Instead of stopping with NUM_READS, stop when the writer thread is done.

Reply:
Done